Create ThriftGemerator

sprsquish opened the pull request:
This is incredibly ugly right now. I'm looking for feedback.

This is a gemerator for thrift gems. It'll create a gem directory structure and all the requisite files to get started on a gem. The thrift files will be ignored in git and regenerated every time the gem is built.

What's nice is, if you want to gemerate your service's thrift files just do the following:

``` scala
class ThriftProject(info: ProjectInfo)
  extends StandardLibraryProject(info)
  with ThriftGemerator
{
  def gemFactory = new ThriftGemFactory(
    "gem-name",
    "GemNameSpace",
    "ThriftService",
    "Some nice description of what this is",
    Seq(("Your Name", "your@email.com")),
    "http://service-homepage.com"
  )
}
```

It also introduces the ability to exclude thrift IDLs from generation. Useful, for instance, when your service presents a scribe endpoint but the generated scribe files shouldn't be a part of the gem.

``` scala
class ThriftProject(info: ProjectInfo)
  extends StandardLibraryProject(info)
  with ThriftGemerator
{
  def gemFactory = new ThriftGemFactory(
    "gem-name",
    "GemNameSpace",
    "ThriftService",
    "Some nice description of what this is",
    Seq(("Your Name", "your@email.com")),
    "http://service-homepage.com",
    Seq("scribe")
  )
}
```

Finally, there are two types of repositories that can be pushed to: Twitter internal and Rubygems. The default is Twitter interal. It can be changed simply by providing it:

``` scala
class ThriftProject(info: ProjectInfo)
  extends StandardLibraryProject(info)
  with ThriftGemerator
{
  def gemFactory = new ThriftGemFactory(
    "gem-name",
    "GemNameSpace",
    "ThriftService",
    "Some nice description of what this is",
    Seq(("Your Name", "your@email.com")),
    "http://service-homepage.com",
    RubygemsGem
  )
}
```

Altogether now:

``` scala
class ThriftProject(info: ProjectInfo)
  extends StandardLibraryProject(info)
  with ThriftGemerator
{
  def gemFactory = new ThriftGemFactory(
    "gem-name",
    "GemNameSpace",
    "ThriftService",
    "Some nice description of what this is",
    Seq(("Your Name", "your@email.com")),
    "http://service-homepage.com",
    Seq("scribe"),
    RubygemsGem
  )
}
```

sprsquish committed:
create ThriftGemerator

robey commented on src/main/scala/com/twitter/sbt/ThriftGemerator.scala:
```diff
@@ -0,0 +1,290 @@
+  val mainTestPath = basePath / "test"
+  val testPath = mainTestPath / name
+  val targetPath = outputPath / "gem"
+  val gemName = name+"-"+version.toString.replaceAll("-SNAPSHOT","")
```
(code style) and line 62, 110, 117, 225, 256, 264

robey commented on src/main/scala/com/twitter/sbt/ThriftGemerator.scala:
```diff
@@ -0,0 +1,290 @@
+  private[this] val uploadTemplate =
+    """require 'rubygems'
+    require 'ods_credentials'
+    = Class.new { extend ODSCredentials }.get_keychain_data.join(':').gsub('!', '\!')
```
what does it mean for a ruby line to begin with "="?

sprsquish replied:
typo. I must've hit undo in my editor before committing.

robey commented on src/main/scala/com/twitter/sbt/ThriftGemerator.scala:
```diff
@@ -0,0 +1,290 @@
+        FileUtilities.write(uploadFile, uploadTemplate, log)
+
+    val exitCode = (("ruby "+uploadFile+" "+(path / gemName).absolutePath) !)
+    if (exitCode == 0) None
```
if (_) {
  ...
} else {
  ...
}

also on line 265

robey commented on src/main/scala/com/twitter/sbt/ThriftGemerator.scala:
```diff
@@ -0,0 +1,290 @@
+  def release(gemName: String, path: Path, log: Logger) = {
+    val uploadFile = (path / "upload.rb").asFile
+    if (!uploadFile.isFile)
+        FileUtilities.write(uploadFile, uploadTemplate, log)
```
braces (or just put it on one line).

robey commented on src/main/scala/com/twitter/sbt/ThriftGemerator.scala:
```diff
@@ -0,0 +1,290 @@
+import _root_.sbt._
+import Process._
+
+class ThriftGemFactory(
```
there are so many ctor params, it's pretty foggy from context what they're used for.

i can think of a couple of ways to fix that:
1. make a case class with good defaults and use that in the ctor, so most people will use the "name=value" form
2. define these as undefined in the class, like "val service: String" so that they're required to be defined when constructed

robey commented:
i'm a big fan of this in general.

sprsquish committed:
cleanup based on feedback

ThriftGemFactory is now a trait with undefined vals
Fixed style issues

sprsquish commented:
Updated based on Robey's feedback. Used a trait with undefined vals for the factory.

``` scala
class ThriftProject(info: ProjectInfo)
  extends StandardLibraryProject(info)
  with ThriftGemerator
{
  def gemFactory = new ThriftGemFactory {
    val name = "gem-name"
    val namespace = "GemNameSpace"
    val service = "ThriftService"
    val description = "Some nice description of what this is"
    val authors = Seq(("Your Name", "your@email.com"))
    val homepage = "http://service-homepage.com"
    override val thriftExclusions = Seq("scribe")
    override val repository = RubygemsGem
  }
}
```

sprsquish committed:
add test-gem task

sprsquish committed:
create a cleanup task for gemerator

sprsquish committed:
Bugfix. ThriftGem was created with missing files

ThriftGem is instantiated before ruby files are generated
so the path was returning an empty set. Use PathFinder instead.

9len commented:
how does this handle different thrift versions? The monorail is still on 0.2.0, but presumably we'd want to publish both 0.2.0 and 0.5.0 versions of things.

sprsquish commented:
There are shims in the monorail so you can provide a thrift 5 gem. The thrift_client version is set as >= 5 which is bound to thrift 2 but wont get in your way if you use thrift_client 6. The downside to that is if you need to use the gem in a service that's bound to thrift 2 but doesn't have the shims. I'd be curious to know if there are other services aside from the monorail that are bound to thrift 2.

sprsquish committed:
use Dir.glob for file list in the gemspec

sprsquish committed:
do not overwrite main include file

sprsquish committed:
run upload through sbt itself

sprsquish committed:
Merge remote-tracking branch 'twitter/master' into create_thrift_gemerator

Conflicts:
	src/main/scala/com/twitter/sbt/CompileThrift.scala

sprsquish commented:
bump on this?
